django/db/backends/mysql/introspection.py

```diff
@@ -67,16 +72,38 @@ def get_table_description(self, cursor, table_name):
         #   not visible length (#5725)
         # - precision and scale (for decimal fields) (#5014)
         # - auto_increment is not available in cursor.description
-        cursor.execute("""
+        info_query = """
             SELECT
                 column_name, data_type, character_maximum_length,
                 numeric_precision, numeric_scale, extra, column_default,
                 CASE
-                    WHEN column_type LIKE '%% unsigned' THEN 1
+                    WHEN column_type LIKE '%%%% unsigned' THEN 1
                     ELSE 0
-                END AS is_unsigned
+                END AS is_unsigned%s
             FROM information_schema.columns
-            WHERE table_name = %s AND table_schema = DATABASE()""", [table_name])
+            WHERE table_name = %%s AND table_schema = DATABASE()
+        """
+        json_query = """
+                ,
+                CASE
+                    WHEN EXISTS (
+                        SELECT 1
+                        FROM information_schema.check_constraints AS c
+                        WHERE c.table_name = %s
+                        AND c.check_clause = 'json_valid(`' + column_name + '`)'
+                        AND c.constraint_name = column_name
+                    )
+                    THEN 1
+                    ELSE 0
+                END AS is_json
+        """
+        if self.connection.features.can_introspect_check_constraints:
+            info_query %= json_query
+            params = [table_name, table_name]
+        else:
+            info_query %= ', CASE WHEN data_type = %s THEN 1 ELSE 0 END AS is_json'
+            params = [FIELD_TYPE.JSON, table_name]
+        cursor.execute(info_query, params)
```

Comment:
This implementation is really confusing to me, and hard to understand, string concatenation is not a good idea to create a query, IMO. I would add an extra query:
<details><summary>diff draft</summary>

```diff
diff --git a/django/db/backends/mysql/introspection.py b/django/db/backends/mysql/introspection.py                                                     
index db1a387679..1b50fde66d 100644                                                                                                                    
--- a/django/db/backends/mysql/introspection.py                                                                                                        
+++ b/django/db/backends/mysql/introspection.py                                                                                                        
@@ -9,7 +9,7 @@ from django.db.backends.base.introspection import (                                                                                    
 from django.db.models.indexes import Index
 from django.utils.datastructures import OrderedSet
 
-FieldInfo = namedtuple('FieldInfo', BaseFieldInfo._fields + ('extra', 'is_unsigned'))
+FieldInfo = namedtuple('FieldInfo', BaseFieldInfo._fields + ('extra', 'is_unsigned', 'has_json_constraint'))
 InfoLine = namedtuple('InfoLine', 'col_name data_type max_len num_prec num_scale extra column_default is_unsigned')
 
 
@@ -35,6 +35,7 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         FIELD_TYPE.MEDIUM_BLOB: 'TextField',
         FIELD_TYPE.LONG_BLOB: 'TextField',
         FIELD_TYPE.VAR_STRING: 'CharField',
+        FIELD_TYPE.JSON: 'JSONField',
     }
 
     def get_field_type(self, data_type, description):
@@ -49,6 +50,10 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
                 return 'PositiveIntegerField'
             elif field_type == 'SmallIntegerField':
                 return 'PositiveSmallIntegerField'
+        # JSON data type is an alias to LONGTEXT on MariaDB, use check
+        # constraints clauses to introspect JSONField.
+        if description.has_json_constraint:
+            return 'JSONField'
         return field_type
 
     def get_table_list(self, cursor):
@@ -62,6 +67,17 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         Return a description of the table with the DB-API cursor.description
         interface."
         """
+        json_constraints = {}
+        if self.connection.features.supports_json_field and self.connection.features.can_introspect_jsonfield:
+            cursor.execute("""
+                SELECT c.constraint_name AS column_name
+                FROM information_schema.check_constraints AS c
+                WHERE
+                    c.table_name = %s AND
+                    LOWER(c.check_clause) = 'json_valid(`' + LOWER(c.constraint_name) + '`)' AND
+                    c.constraint_schema = DATABASE()
+            """, [table_name])
+            json_constraints = {column_name for column_name in cursor.fetchall()}
         # information_schema database gives more accurate results for some figures:
         # - varchar length returned by cursor.description is an internal length,
         #   not visible length (#5725)
@@ -96,6 +112,7 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
                 info.column_default,
                 info.extra,
                 info.is_unsigned,
+                line[0] in json_constraints,
             ))
         return fields
 
                                                                                                                                
```
</details>

so we can check constraints on MariaDB and check data type on MySQL.

Reply:
That makes much more sense. Thanks!